### PR TITLE
Optimize matrix scanning

### PR DESCRIPTION
**Pull request by eltang**
I realized that "optimizations" is too general a name for a branch. Here is the description of #342.

> The optimizations in this PR are much more complete than those in #278.
> 
> @jackhumbert I'm not able to test the code that accommodates backwards diodes, though I see no reason why it shouldn't work. The rest, however, is working well on my Planck. Let me know if you have any questions.

@jackhumbert How's this?

- Optimize all loops to count down where practical
- Rename the `COLS`, `ROWS`, and `DEBOUNCE` macros for clarity
- Add the `UNUSED_PINS` macro (it's good to set them unused pins as input with pull-up so that they don't waste power constantly drifting between states, but I'm not sure if `matrix.c` would be the right place to do this)
- Store the pin configuration as an array of `io_pin_t` variables to allow the addresses of I/O registers to be calculated directly
- Make the debouncing code wait for the proper amount of time (1 ms more than before)
- Clean up Bluetooth configuration code
- Ensure that no lines are wider than 80 characters

### CODE REVIEW

**Comment by jackhumbert** [comment]
This is looking good! A little feedback:

* The branch name is irrelevant when making pull requests (no need to make more than one PR in this case).
* If you close/open PRs, please say what you're doing - it makes it look like things aren't ready. This goes the same for removing things (eg the ErgoDox parts).
* Can you describe (edit your main description) the changes besides "optimizing matrix scanning"? It's useful to link back to old things, but don't assume people will dig up the entire history. Here's a quick example of explaining more:

> * Renamed `COLS` and `ROWS` for clarity
> * Added `UNUSED_PINS ` for ... (not entirely sure here)

On that, we can guess why you did things, but ultimately we won't know unless you explain them. Comments in the code can help a lot with that - correct me if I'm wrong, but it doesn't look like you added any of your own, and in some cases removed existing ones.

There's some wacky diff stuff going on with the `matrix.c` (not your fault), so if you could describe the changes you've made there, that would be awesome. It looks like you might have changed some of the debouncing stuff, but that could be the diff messing things up - otherwise, explaining that would be nice!

**Comment by eltang** [comment]
Sorry, I'm obviously not good with PRs yet.

File: quantum/matrix.c [hunk 1]
```diff
@@ -15,300 +15,209 @@ GNU General Public License for more details.
...
-}
-
-uint8_t matrix_key_count(void)
-{
-    uint8_t count = 0;
-    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
-        count += bitpop16(matrix[i]);
+    if (debouncing_delay >= 0) {
+        dprintf("Debouncing delay remaining: %X\n", debouncing_delay);
+        if (--debouncing_delay >= 0) {
```
Comment (jackhumbert) [inline quantum/matrix.c]: As clean as this is, it's not as readable as separating the actions.

**Comment by jackhumbert** [comment]
What's your reasoning for making the loops count down?

On the debouncing code, what makes your changes the proper amount of time?

Would you mind updating [`/HAND_WIRE.md`](https://github.com/jackhumbert/qmk_firmware/blob/master/HAND_WIRE.md#configh) as well?

File: quantum/matrix.c (hunk 1 above)
Comment (eltang) [inline quantum/matrix.c]: I'm not sure I understand this sentence.

Comment (jackhumbert) [inline quantum/matrix.c]: Always prioritise readability over elegance - anyone should be able to quickly look at the code and figure out what's happening. Comments can help a lot for more complicated actions, but splitting up the `debouncing_delay--` and the conditional is the easier way to go here.

**Comment by eltang** [comment]
See [this](http://www.atmel.com/images/doc8453.pdf). I think the rest of the code could really benefit from these optimizations.

Right now, debouncing stops when the counter variable reaches `0`. I'm simply allowing it to continue until it reaches `-1`.

Done.

Should I add some code to set up the unused pins to this PR? If so, to which file?

**Comment by jackhumbert** [comment]
On the loop decrementing, it looks like that's only saving 2 bytes when setting the condition to the counter (`r`), not keeping the `r >= 0` - either way, it's not a big save. I don't really care if you write new loops like that, but please don't go around changing them.

> Right now, debouncing stops when the counter variable reaches 0. I'm simply allowing it to continue until it reaches -1.

But why?

> Should I add some code to set up the unused pins to this PR? If so, to which file?

If you'd like to - let's escape it with a `#ifdef TURN_OFF_UNUSED_PINS` or something defaulted to off, though. I would imagine there's code out there currently that assumes they're set to the defaults.

**Comment by eltang** [comment]
If I had proper knowledge of asm I could know for sure, but I'm pretty sure that it works this way too. But okay.

I'm allowing the code to wait for one more millisecond. Right now, if `debounce` is set to `5`, then the code decrements the variable once per every call of `matrix_scan` In theory, it should wait for five milliseconds, but it really waits for four—on the final time, `debounce` is checked to be zero, and the code branches away from `wait_ms` prematurely. Now, I'm making it continue if the counter is `0` so that the debouncing can continue for one more millisecond. Does that make sense?

I guess I'll leave those definitions blank for now then. This might be something for another PR.

**Comment by jackhumbert** [comment]
Might be something you could test ;)

Ah, gotcha. That sounds good - this will make a value of `0` behave correctly, right?

Sounds good - I like the idea of listing the unused pins for reference as well (ones unused by the matrix mapping) for modding.

**Comment by eltang** [comment]
Well, I tried to look at the assembling listings for the old and new versions of the file, but I forgot that the old file was a lot different. The ratio of comparison instructions ended up being 54:5, which is good, but not useful for my comparison.

Exactly.

It's not causing any damage to the hardware as far as I know, but making code to turn on the pull-up resistors would be beneficial for saving power.

Is this PR ready then?

**Comment by eltang** [comment]
Actually, I could test using one of the other files. I'll be back.

**Comment by jackhumbert** [comment]
It'll be more straight-forward to just change that one line - unless you're talking about all of your changes (I wasn't). 

Sounds like it! Pulling to test myself now.

**Comment by eltang** [comment]
I'm dumb.

You will likely experience a decrease in typos. :D I think it will especially benefit Colemak users like you.

**Comment by jackhumbert** [comment]
Merged! I'm not sure how this will decrease typos, though - what have you experienced?

**Comment by eltang** [comment]
Thanks! Well, at one stage, I changed the `_delay_us(30);` line to `wait_ms(1)`, and it seemed like I was transposing a lot more letters than usual. I eventually got rid of the line completely, because it made absolutely no difference, and the typos went away.

**Comment by jackhumbert** [comment]
That's probably a good example of something not to change - we can keep it in for now, but I would imagine it's there for a reason. Maybe your asm stuff counters it. Let's keep an eye out for problems related to that, though.

**Comment by eltang** [comment]
Well, the purpose was to wait for the reading to stabilize, but in my tests, it wasn't needed—I had debugging messages on, and none were produced when I wasn't typing. I can write asm, but only in my dreams.

**Comment by jackhumbert** [comment]
I was referring to the `_SFR_IO8` stuff - you were able to test it on your board, but not all boards/processors.

**Comment by eltang** [comment]
That's just a macro. If anything, it makes things faster. Don't worry, the macro works on all AVRs.

**Comment by jackhumbert** [comment]
That's what I mean - maybe the wait isn't needed with your set/read implementation. We'll see, though.

**Comment by eltang** [comment]
So my guess was correct. I changed one of the loops to count up and there were two more comparison instructions. The diff is (https://gist.github.com/eltang/0ba66a4bb6e9de4f4ab453fae24507b1/revisions)[here], if you're interested. The regex to search `(cpse|cp|cpc|cpi)`.

**Comment by jackhumbert** [comment]
So does this actually affect the size of the hex? It's not super clear from the .lst.

**Comment by eltang** [comment]
It does. All of the asm makes it into the .hex file. Have you changed your mind about optimizing all the loops?

**Comment by eltang** [comment]
I just checked and that single change increases the size eight bytes.

File: quantum/matrix.c (hunk 1 above)
Comment (eltang) [inline quantum/matrix.c]: Sorry, GitHub has been acting up and I'm just seeing this. I didn't write the original condition, but I get what you're saying. I do strive to make my code clean enough not to require comments, though.

**Comment by jackhumbert** [comment]
Nice - I'll have to mess around with that myself.

**Comment by eltang** [comment]
Be sure to use the prefix decrement operator then.

**Comment by jackhumbert** [comment]
Were you able to test that change as well?

**Comment by eltang** [comment]
It's in the document of tips that I gave you.

**Comment by jackhumbert** [comment]
Yeah, I'm asking if you were able to test the size differences in this application.

**Comment by eltang** [comment]
There was no difference this time, but I suspect that the compiler was able to optimize things. Anyway, it's still good to use the prefix decrement and prefix increment operators whenever possible (when the previous value will not be used) because not all compilers are as smart.

**Comment by jackhumbert** [comment]
That was my suspicion as well. I would imagine everyone is using the same compiler here (avr-gcc), so I'm not sure it's something we really need to mess with.

When you edit your comments, put a "edit" or something in it - it breaks the conversation otherwise.

**Comment by eltang** [comment]
Well, WinAVR does use an older version of the compiler, so that might be a factor.

Will do.

**Comment by jackhumbert** [comment]
The prefix/postfix optimisation is a pretty basic one - I'd be surprised if any modern compiler didn't have it. If you find anything else about it, feel free to post it here.

**Comment by eltang** [comment]
Actually, looking at the size of the .hex file is not a good way to gauge the final amount of flash that will be taken up on the chip. The proper way to do that is to run `avr-size`. The command to run if the firmware is for a Planck, for example, would be `avr-size -C --mcu=atmega32u4 planck.elf`.

**Comment by jackhumbert** [comment]
That's what the rules.mk file outputs, which is what I was referring to, not the literal size of the .hex file on your HD. Check out that file for commands that work (yours does not).

**Comment by eltang** [comment]
Really? The command works just fine for me when I'm in the folder where the .elf file is. It even gives me the percentage of the flash on the chip that will be used (like `dfu-programmer`).

**Comment by fredizzimo** [comment]
I'm not sure it was a smart idea to merge this pull request, and I would like to suggest that we revert the matrix scanning code back to the previous version. 

The major problem I have, is that the matrix is read immediately after setting the pins
```
toggle_row(r);
matrix_row_t state = read_cols();
```

That will most likely cause problems, since it takes some time for the electric signal to go through the keyboard. I personally don't have the knowledge, but if you look at Hasu's code [here] https://github.com/tmk/tmk_keyboard/blob/master/keyboard/hhkb/matrix.c), for example, it seems like things are really sensitive. The original comment also indicates that a delay is needed `"_delay_us(30);  // without this wait read unstable value.` 

I'm pretty sure that Hasu has done a lot of testing with this, as can be seen [here](https://geekhack.org/index.php?topic=42385.0). That thread also shows that he is aware that the code only waits for 4ms rather than 5.

I suspect that the debouncing code masks the problem, since eventually there will be 5 reads in the row that reads correctly. One way to test if it really works, would be to disable the debouncing and hold a key down, if it constantly reports that it's held down, then it might work correctly without a delay on that particular keyboard, but if it sometimes reports that the key is up, then there are problems. In any case I can't see how a 30 us delay per row/col, could have any effect on the accuracy of the typing.

Also the UNUSED_PINS define is still not used anywhere in the code, and there's some misleading documentation that tells you that it saves power.

**Comment by jackhumbert** [comment]
That sounds good - getting this merged/to the point of being merged was a headache anyway. 

Would you be comfortable making a PR with the revert? I'm not entirely sure how to do that outside of just looking back and copying code. I am depending on some of the macro definitions elsewhere, and I think the `_SFR_IO8` format could probably stay/worked into better, readable macros.

**Comment by fredizzimo** [comment]
A `git revert --no-commit`, followed by using a diff tool for selectively selecting the changes that we want to keep, would probably be the best way to go.

I can do that at some point, since I want to add some abstractions(functionally equivalent) afterwards, in order to enable ChibiOS and AVR, to share most of the matrix scanning code, so that we don't need to duplicate the code. I also don't like that there's two different code paths for reverse and forward diode directions, the could could most likely be slightly refactored to share most of the code.

I'm confident in doing the actual code modifications, but testing is the biggest problem. I will make sure that I have my Chibios based Ergodox running at that point, so that I at least can test that path. But we also need to test with at least on AVR based keyboard. I think just a quick testing should be enough, since we should be able to quite easily do a code review to verify that the new code is working the same way as the old one.

I didn't comment about the code size optimizations yesterday, since I didn't have time to investigate. But they are wrong too, first read the documentation carefully

> Loops are widely used in 8-bit AVR code. There are “while ( ) { }” loop, “for ( )” loop
and “do { } while ( )” loop. If the -Os optimization option is enabled; the compiler will
optimize the loops automatically to have the same code size.
>
> However we can still reduce the code size slightly. If we use a **“do { } while ( )” loop**,
an increment or a decrement loop index generates different code size. Usually we
write our loops counting from zero to the maximum value (increment), but it is more
efficient to count the loop from the maximum value to zero (decrement).
That is because in an increment loop, a comparison instruction is needed to compare
the loop index with the maximum value in every loop to check if the loop index
reaches the maximum value.
>
>When we use a decrement loop, this comparison is not needed any more because
the decremented result of the loop index will set the Z (zero) flag in SREG if it
reaches zero. 

You can first see that the compiler automatically optimizes any for loop. And to take advantage of the two byte optimization, you need to use a do while loop, where the comparison can be to exactly zero (not <= 0).

We can easily verify that the optimization hasn't been applied by looking at assembly output from the matrix_key_count function for example.
```
267:../../quantum/matrix.c **** uint8_t matrix_key_count(void) {
 638               		.loc 1 267 0
 639               		.cfi_startproc
 640               		.loc 1 267 0
 641 0000 0F93      		push r16
 642               	.LCFI12:
 643               		.cfi_def_cfa_offset 3
 644               		.cfi_offset 16, -2
 645 0002 1F93      		push r17
 646               	.LCFI13:
 647               		.cfi_def_cfa_offset 4
 648               		.cfi_offset 17, -3
 649 0004 CF93      		push r28
 650               	.LCFI14:
 651               		.cfi_def_cfa_offset 5
 652               		.cfi_offset 28, -4
 653               	/* prologue: function */
 654               	/* frame size = 0 */
 655               	/* stack size = 3 */
 656               	.L__stack_usage = 3
 657               	.LVL34:
 658 0006 00E0      		ldi r16,lo8(matrix+10)
 659 0008 10E0      		ldi r17,hi8(matrix+10)
 268:../../quantum/matrix.c ****     uint8_t count = 0;
 660               		.loc 1 268 0
 661 000a C0E0      		ldi r28,0
 662               	.LVL35:
 663               	.L35:
 664               	.LBB35:
 269:../../quantum/matrix.c ****     for (int8_t r = MATRIX_ROWS - 1; r >= 0; --r) {
 270:../../quantum/matrix.c ****         count += bitpop16(matrix_get_row(r));
 665               		.loc 1 270 0 discriminator 3
 666 000c F801      		movw r30,r16
 667 000e 9291      		ld r25,-Z
 668 0010 8291      		ld r24,-Z
 669 0012 8F01      		movw r16,r30
 670 0014 0E94 0000 		call bitpop16
 671               	.LVL36:
 672 0018 C80F      		add r28,r24
 673               	.LVL37:
 269:../../quantum/matrix.c ****     for (int8_t r = MATRIX_ROWS - 1; r >= 0; --r) {
 674               		.loc 1 269 0 discriminator 3
 675 001a F0E0      		ldi r31,hi8(matrix)
 676 001c 0030      		cpi r16,lo8(matrix)
 677 001e 1F07      		cpc r17,r31
 678 0020 01F4      		brne .L35
 679               	.LBE35:
 271:../../quantum/matrix.c ****     }
 272:../../quantum/matrix.c ****     return count;
 273:../../quantum/matrix.c **** }
 680               		.loc 1 273 0
 681 0022 8C2F      		mov r24,r28
 682               	/* epilogue start */
 683 0024 CF91      		pop r28
 684               	.LVL38:
 685 0026 1F91      		pop r17
 686 0028 0F91      		pop r16
 687 002a 0895      		ret
 688               		.cfi_endproc
 689               	.LFE22:
```
The interesting part is here
```
 675 001a F0E0      		ldi r31,hi8(matrix)
 676 001c 0030      		cpi r16,lo8(matrix)
 677 001e 1F07      		cpc r17,r31
 678 0020 01F4      		brne .L35
```

As you can see there's no less than two comparison functions("compare with intermediate, and "compare with carry", while if the optimization works properly, we should have none. The two comparisons comes from the fact that the compiler is smart, really smart. It has completely eliminated the index, and operates on pointers instead, therefore eliminating the multiply that would be needed for each loop step.

Changing the code back to the original form, yields the same optimization, but the memory is traversed in the forward order, rather than backwards. The code size remains the same.

We can't do any better than this, for loops of this kind. However some loops could be transformed to a while loop form as described in the document. However for almost all of those loops it would be trivial for the compiler to apply the same optimization. So I always recommend checking the output before changing the code to something less readable.

So with that conclusion, I think we should remove the loop "optimizations" as well.

**Pull request merged**